File: android/app/src/main/java/com/fitsense/ai/device/SupportedDevices.kt
package com.fitsense.ai.device

import android.content.Context
import org.json.JSONObject

/**
 * Limited launch allowlist — assets/supported-devices.json.
 * Uncertified devices may still measure; UI should warn, never invent mm.
 */
object SupportedDevices {
    data class Evaluation(
        val status: String,
        val launchLimited: Boolean,
        val message: String,
    )

    fun evaluate(context: Context, manufacturer: String?, model: String?): Evaluation {
        val mfr = manufacturer?.trim().orEmpty().lowercase()
        val mdl = model?.trim().orEmpty().lowercase()
        val json = runCatching {
            context.assets.open("supported-devices.json").bufferedReader().use { it.readText() }
        }.getOrNull() ?: return Evaluation(
            status = "unknown",
            launchLimited = true,
            message = "Supported-device list missing — treat sizing as limited until certified.",
        )
        val root = JSONObject(json)
        val devices = root.optJSONArray("devices") ?: return Evaluation(
            "unknown",
            true,
            "No devices listed — limited launch.",
        )
        for (i in 0 until devices.length()) {
            val entry = devices.getJSONObject(i)
            val entryMfr = entry.optString("manufacturer").lowercase()
            val mfrOk = mfr.isEmpty() || mfr.contains(entryMfr) || entryMfr.contains(mfr)
            if (!mfrOk) continue
            val tokens = entry.optJSONArray("modelContains") ?: continue
            var hit = mdl.isEmpty()
            for (t in 0 until tokens.length()) {
                if (mdl.contains(tokens.getString(t).lowercase())) {
                    hit = true
                    break
                }
            }
            if (!hit) continue
            return when (entry.optString("status")) {
                "certified" -> Evaluation(
                    "certified",
                    false,
                    "Device cohort is certified for retail millimetre measurement.",
                )
                "blocked" -> Evaluation(
                    "blocked",
                    true,
                    "This device is blocked for retail sizing.",
                )
                else -> Evaluation(
                    "candidate",
                    true,
                    "Launch candidate — certify via Brannock study before treating mm as sizing truth.",
                )
            }
        }
        return Evaluation(
            "unknown",
            true,
            "Device is outside the limited supported list — label results as estimate until certified.",
        )
    }
}

File: src/lib/supportedDevices.ts
/**
 * Limited launch device allowlist — mirrors docs/supported-devices.json.
 * Uncertified devices may still measure; UI should warn, never invent mm.
 */
import supported from "../../docs/supported-devices.json";

export type DeviceStatus = "candidate" | "certified" | "blocked";

export type SupportedDeviceEntry = {
  manufacturer: string;
  modelContains: string[];
  status: DeviceStatus;
  region?: string;
  notes?: string;
};

export type DeviceCertification = {
  status: DeviceStatus | "unknown";
  entry: SupportedDeviceEntry | null;
  launchLimited: boolean;
  message: string;
};

const catalog = supported as {
  minAndroidApi?: number;
  devices: SupportedDeviceEntry[];
};

export function evaluateDeviceCertification(
  manufacturer?: string | null,
  model?: string | null,
): DeviceCertification {
  const mfr = (manufacturer ?? "").trim().toLowerCase();
  const mdl = (model ?? "").trim().toLowerCase();
  if (!mfr && !mdl) {
    return {
      status: "unknown",
      entry: null,
      launchLimited: true,
      message:
        "Device not identified — retail sizing is limited until this phone is certified.",
    };
  }

  for (const entry of catalog.devices) {
    const mfrOk =
      !mfr ||
      entry.manufacturer.toLowerCase() === mfr ||
      mfr.includes(entry.manufacturer.toLowerCase());
    if (!mfrOk) continue;
    const hit = entry.modelContains.some((token) =>
      mdl.includes(token.toLowerCase()),
    );
    if (!hit && mdl) continue;
    if (entry.status === "blocked") {
      return {
        status: "blocked",
        entry,
        launchLimited: true,
        message:
          "This device is blocked for retail sizing. Use a certified phone or continue for estimate-only demo.",
      };
    }
    if (entry.status === "certified") {
      return {
        status: "certified",
        entry,
        launchLimited: false,
        message: "Device cohort is certified for retail millimetre measurement.",
      };
    }
    return {
      status: "candidate",
      entry,
      launchLimited: true,
      message:
        "Device is on the launch candidate list — certify via Brannock study before treating mm as sizing truth.",
    };
  }

  return {
    status: "unknown",
    entry: null,
    launchLimited: true,
    message:
      "Device is outside the limited supported list. Measurement may continue; label results as estimate until certified.",
  };
}

export function supportedDeviceCatalog(): SupportedDeviceEntry[] {
  return catalog.devices;
}
